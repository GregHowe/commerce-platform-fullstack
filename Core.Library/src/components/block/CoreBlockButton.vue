<template>
	<div class="block__body">
		<a
			v-if="!isInternal"
			role="button"
			:href="buttonUrl"
			:target="newTabTarget"
			class="button"
			@click="clickEvent"
		>
			<span class="button_icon--custom">
				<CoreIcon
					v-if="settings.icon"
					:icon="settings.icon"
					name="buttonIcon"
					class="button_icon"
				/>
			</span>
			<span class="button_label button_icon--down">
				<span class="button_label_text">{{ buttonLabel }}</span>
				<CoreIcon
					icon="down-arrow"
					class="button_icon button_i"
				/>
			</span>
		</a>
		<NuxtLink
			v-else
			role="button"
			:to="buttonUrl"
			:target="newTabTarget"
			class="button"
			@click="clickEvent"
		>
			<span class="button_icon--custom">
				<CoreIcon
					v-if="settings.icon"
					:icon="settings.icon"
					name="buttonIcon"
					class="button_icon"
				/>
			</span>
			<span class="button_label button_icon--down">
				<span class="button_label_text">{{ buttonLabel }}</span>
				<CoreIcon
					icon="down-arrow"
					class="button_icon button_i"
				/>
			</span>
		</NuxtLink>
	</div>
</template>

<script>
export default {
	name: "CoreBlockButton",
	props: {
		settings: {
			type: Object,
			required: true,
		},
	},
	computed: {
		newTabTarget() {
			return this.settings.newTab ? "_blank" : null;
		},
		buttonUrl() {
			return this.settings?.url || null;
		},
		buttonLabel() {
			return this.settings?.label || "Button Label";
		},
		clickEvent() {
			return this.settings?.event || this.noEventFound;
		},
		isInternal() {
			return this.settings?.url?.charAt(0) == "/";
		},
	},
	methods: {
		noEventFound() {},
	},
};
</script>

<style lang="scss">
.block_button {
	// padding: 0.125rem; /* spacing */
	font-family: var(--_core__button_font-family);
	.block__body {
		width: fit-content;
		white-space: nowrap;
	}

	&,
	&.block_button_primary {
		.button {
			cursor: pointer;
			padding: 0.75rem 1.25rem;
			display: flex;
			flex-direction: row;
			background-color: var(--core__color_primary);
			border: 1px solid var(--core__color_primary);
			border-radius: var(--_core__button-primary_border-radius);
			color: var(--core__color_white);
			text-decoration: none;
		}
		.button_label {
			display: flex;
			align-items: center;
		}
		&.block_button_outline {
			.button {
				background-color: transparent;
				color: var(--core__color_primary);
			}
		}
		&.block_button_invert {
			.button {
				background-color: var(--core__color_white);
				border-color: var(--core__color_white);
				color: var(--core__color_primary);
			}
			&.block_button_outline {
				.button {
					background-color: transparent;
					border-color: var(--core__color_white);
					color: var(--core__color_white);
				}
			}
		}
	}

	&.block_button_secondary {
		.button {
			background-color: var(--core__color_white);
			border-color: var(--core__color_primary);
			color: var(--core__color_primary);
		}
		&.block_button_outline {
			.button {
				background-color: transparent;
				border-color: var(--core__color_primary);
				color: var(--core__color_primary);
			}
		}
		&.block_button_invert {
			.button {
				background-color: var(--core__color_primary);
				border-color: var(--core__color_primary);
				color: var(--core__color_primary);
			}
			&.block_button_outline {
				.button {
					background-color: #ff0000;
					border-color: var(--core__color_primary);
					color: var(--core__color_primary);
				}
			}
		}
	}

	&.block_button_tertiary {
		.button {
			background: transparent;
			border-color: transparent;
			color: var(--core__color_primary);
			padding-left: 0;
			font-size: 18px;

			.button_icon--custom svg {
				display: flex;
			}
		}
	}

	&.block_button_link {
		.button {
			background: transparent;
			border-color: transparent;
			color: var(--_core__link_color);
			text-decoration: var(--_core__link_text-decoration);
			padding: 0;
		}
	}

	&.block_button_text {
		.button {
			background: transparent;
			border-color: transparent;
			color: var(--core__color_primary);
			padding-left: 0;
			font-weight: 700;
			font-size: 18px;

			.button_icon--custom svg {
				display: flex;
			}
		}
	}

	&.block_button_align-left {
		.block__body {
			margin-left: 0;
		}
	}

	&.block_button_align-right {
		.block__body {
			margin: 0 auto;
		}
	}

	&.block_button_align-right {
		.block__body {
			margin-right: 0;
		}
	}

	.button_icon--custom svg {
		margin: 0 1rem 0 -0.5rem;
	}

	.button_icon--down svg {
		display: none;
	}

	&.block_button_scroll {
		.button {
			background: transparent;
			border-color: transparent;
			color: var(--_core__button_primary-background);
			align-items: center;
		}
		.button_label {
			flex-direction: column;
		}
		.button_icon--custom svg {
			display: inline;
		}
		.button_icon--down svg {
			display: block;
			margin: -0.25rem auto -1.25rem;
		}
	}

	&.block_button_icon-right {
		.button {
			flex-direction: row-reverse;
			.button_icon--custom svg {
				margin: 0 -0.5rem 0 1rem;
			}
		}
	}
}
</style>
