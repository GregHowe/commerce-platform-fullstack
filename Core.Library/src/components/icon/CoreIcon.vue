<template>
	<component
		:is="iconName"
		v-bind="$attrs"
		:value="icon"
	/>
</template>

<script>
import {
	kebabCase as _kebabCase,
	camelCase as _camelCase,
	upperFirst as _upperFirst,
	flow as _flow,
} from "lodash";
import iconSchema from "../../schemas/icon.json";
export default {
	name: "CoreIcon",
	props: {
		icon: {
			type: String,
			required: true,
		},
	},
	computed: {
		iconName() {
			return iconSchema.some(
				(icon) => icon.value === _kebabCase(this.icon)
			)
				? "CoreIcon" + _flow(_camelCase, _upperFirst)(this.icon)
				: "CoreIconQuestionMark";
		},
	},
};
</script>

<style></style>
