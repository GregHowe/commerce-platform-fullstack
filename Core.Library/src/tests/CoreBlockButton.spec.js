import CoreBlockButton from "../components/block/CoreBlockButton";
import CoreIcon from "../components/icon/CoreIcon";
import CoreIconLeftArrow from "../components/icon/CoreIconLeftArrow";
import CoreIconDownArrow from "../components/icon/CoreIconDownArrow";
import "@testing-library/jest-dom";
import { render } from "@testing-library/vue";

const renderComponent = async (props = {}) => {
	return render(CoreBlockButton, {
		props,
		components: { CoreIcon, CoreIconLeftArrow, CoreIconDownArrow },
	});
};
describe("CoreBlockButton", () => {
	test("left arrow icon is shown is shown if passed into prop", async () => {
		const { container } = await renderComponent({
			settings: {
				icon: "left-arrow",
			},
		});
		expect(container).toMatchSnapshot();
	});

	test("if url is set, make sure button href matches", async () => {
		const { container } = await renderComponent({
			settings: { url: "https://www.google.com" },
		});
		expect(container).toMatchSnapshot();
	});

	test("if newTab is set, make sure button target matches", async () => {
		const { container } = await renderComponent({
			settings: { newTab: "_blank" },
		});
		expect(container).toMatchSnapshot();
	});

	test("if url is an internal link, nuxt link is used for routing", async () => {
		const { container } = await renderComponent({
			settings: { url: "/internal-link" },
		});
		expect(container).toMatchSnapshot();
	});
});
