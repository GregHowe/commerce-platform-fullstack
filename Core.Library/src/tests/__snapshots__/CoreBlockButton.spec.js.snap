// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreBlockButton if newTab is set, make sure button target matches 1`] = `
<div>
  <div
    class="block__body"
  >
    <a
      class="button"
      role="button"
      target="_blank"
    >
      <span
        class="button_icon--custom"
      >
        <!---->
      </span>
       
      <span
        class="button_label button_icon--down"
      >
        <span
          class="button_label_text"
        >
          Button Label
        </span>
         
        <span
          class="button_icon button_i"
          icon="down-arrow"
        />
      </span>
    </a>
  </div>
</div>
`;

exports[`CoreBlockButton if url is an internal link, nuxt link is used for routing 1`] = `
<div>
  <div
    class="block__body"
  >
    <a
      class="button"
      role="button"
      test-id="nuxtlink-stub"
      to="/internal-link"
    >
      <span
        class="button_icon--custom"
      >
        <!---->
      </span>
       
      <span
        class="button_label button_icon--down"
      >
        <span
          class="button_label_text"
        >
          Button Label
        </span>
         
        <span
          class="button_icon button_i"
          icon="down-arrow"
        />
      </span>
    </a>
  </div>
</div>
`;

exports[`CoreBlockButton if url is set, make sure button href matches 1`] = `
<div>
  <div
    class="block__body"
  >
    <a
      class="button"
      href="https://www.google.com"
      role="button"
    >
      <span
        class="button_icon--custom"
      >
        <!---->
      </span>
       
      <span
        class="button_label button_icon--down"
      >
        <span
          class="button_label_text"
        >
          Button Label
        </span>
         
        <span
          class="button_icon button_i"
          icon="down-arrow"
        />
      </span>
    </a>
  </div>
</div>
`;

exports[`CoreBlockButton left arrow icon is shown is shown if passed into prop 1`] = `
<div>
  <div
    class="block__body"
  >
    <a
      class="button"
      role="button"
    >
      <span
        class="button_icon--custom"
      >
        <span
          class="button_icon"
          icon="left-arrow"
          name="buttonIcon"
        />
      </span>
       
      <span
        class="button_label button_icon--down"
      >
        <span
          class="button_label_text"
        >
          Button Label
        </span>
         
        <span
          class="button_icon button_i"
          icon="down-arrow"
        />
      </span>
    </a>
  </div>
</div>
`;
